src/mutex.cpp

```diff
@@ -99,8 +111,14 @@ DmtcpMutexUnlock(DmtcpMutex *mutex)
   if (mutex->count == 0) {
     mutex->owner = 0;
 
-    // TODO(Kapil): Call futex_wake only if there are futex waiters.
-    JASSERT(futex_wake((uint32_t*) &mutex->owner, 1) != -1) (JASSERT_ERRNO);
+    if (__sync_bool_compare_and_swap(&mutex->futex, LOCK_ACQUIRED, LOCK_FREE)) {
+      // No need to call futex_wake as we don't have any waiters.
+    } else {
+      JASSERT(__sync_bool_compare_and_swap(&mutex->futex,
+                                           LOCK_ACQUIRED_WAITERS_QUEUED,
+                                           LOCK_FREE));
```

Comment:
QUESTION:  Suppose we succeed in setting the lock to LOCK_FREE, above.  Then two other threads come in.  One acquires the lock and sets the state to LOCK_ACQUIRED.  A second thread comes in and sets the state to LOCK_ACQUIRED_WAITERS_QUEUED and does `futex_wait()`.  It appears that we are now going to send a _spurious_ wakeup to the thread that is waiting on the lock.
&nbsp;&nbsp; That's okay, if your intention is that the thread receiving the spurious wakeup will simply go around the loop, and return to waiting.  But in that case, it would be nice to add a comment here, such as:
```
// After setting the lock to LOCK_FREE, two other threads may have tried to grab the lock.
// One succeeds and one is waiting.  In this case, we are doing a spurious wakeup.
// The logic for handling a spurious wakeup is contained in the function 'DmtcpMutexLock()'.

Reply:
And I'll add one further comment on the simplicity of a rslock with readers preferred.  This is a quote from:
`glibc-2.32/nptl/pthread_rwlock_common.c`:
>   This lock is built so that workloads of mostly readers can be executed with           low runtime overheads.  This matches that the default kind of the lock is             PTHREAD_RWLOCK_PREFER_READER_NP.  Acquiring a read lock requires a single             atomic addition if the lock is or was previously acquired by other                    readers; releasing the lock is a single CAS if there are no concurrent                writers.

And of course, `CAS`, above, stands for Compare-And-Swap.

Reply:
The comment for lock is sufficient. We won't have spurious wakeups, so no need for this comment here.